server/validate.go

```diff
@@ -106,47 +106,62 @@ func validateGetEntryAndProofRequest(req *trillian.GetEntryAndProofRequest) erro
 }
 
 func validateQueueLeafRequest(req *trillian.QueueLeafRequest) error {
-	if req.Leaf == nil {
-		return status.Error(codes.InvalidArgument, "QueueLeafRequest.Leaf empty")
-	}
-	if err := validateLogLeaf(req.Leaf); err != nil {
-		// validateLogLeaf errors are meant to chain nicely with "Leaf."
-		return status.Errorf(codes.InvalidArgument, "QueueLeafRequest.Leaf.%v", err)
-	}
-	return nil
+	return validateLogLeaf(req.Leaf, "QueueLeafRequest.Leaf")
+}
+
+func validateAddSequencedLeafRequest(req *trillian.AddSequencedLeafRequest) error {
+	return validateLogLeaf(req.Leaf, "AddSequencedLeafRequest.Leaf")
 }
 
 func validateQueueLeavesRequest(req *trillian.QueueLeavesRequest) error {
-	if len(req.Leaves) == 0 {
-		return status.Error(codes.InvalidArgument, "QueueLeavesRequest.Leaves empty")
+	return validateLogLeaves(req.Leaves, "QueueLeavesRequest")
+}
+
+func validateAddSequencedLeavesRequest(req *trillian.AddSequencedLeavesRequest) error {
+	prefix := "AddSequencedLeavesRequest"
+	if err := validateLogLeaves(req.Leaves, prefix); err != nil {
+		return err
 	}
+
+	// Note: Not empty, as verified by validateLogLeaves.
+	nextIndex := req.Leaves[0].LeafIndex
 	for i, leaf := range req.Leaves {
-		if leaf == nil {
-			return status.Errorf(codes.InvalidArgument, "QueueLeavesRequest.Leaves[%v] empty", i)
-		}
-		if err := validateLogLeaf(leaf); err != nil {
-			// validateLogLeaf errors are meant to chain nicely with "Leaves."
-			return status.Errorf(codes.InvalidArgument, "QueueLeavesRequest.Leaves[%v].%v", i, err)
+		if leaf.LeafIndex != nextIndex {
+			return status.Errorf(codes.OutOfRange, "%v.Leaves[%v].LeafIndex=%v, want %v", prefix, i, leaf.LeafIndex, nextIndex)
```

Comment:
This is checking for sequential index? Are we using OutOfRange to be able to distinguish it because it could be FailedPrecondition?

Reply:
Right, let's make it a `FailedPrecondition`.